### PR TITLE
Skip excess property elaborations

### LINKED ISSUE #29759: Error messages should place most relevant information first.
Here's an actual example of an error message that can go wrong in our library when mispelling a property when creating an object literal:

```
[ts]
Type '{ containers: { nginx: { image: string; memory: number; portMappings: NetworkListener[]; enviroment: number; }; }; }' is not assignable to type 'FargateTaskDefinitionArgs'.
  Types of property 'containers' are incompatible.
    Type '{ nginx: { image: string; memory: number; portMappings: NetworkListener[]; enviroment: number; }; }' is not assignable to type 'Record<string, Container>'.
      Property 'nginx' is incompatible with index signature.
        Type '{ image: string; memory: number; portMappings: NetworkListener[]; enviroment: number; }' is not assignable to type 'Container'.
          Object literal may only specify known properties, but 'enviroment' does not exist in type 'Container'. Did you mean to write 'environment'? [2322]

- fargateService.d.ts(169, 5): The expected type comes from property 'taskDefinitionArgs' which is declared here on type 'FargateServiceArgs'
```

The funny thing here is that the final parts of the error message are superb. Namely:

```
- Object literal may only specify known properties, but 'enviroment' does not exist in type 'Container'. Did you mean to write 'environment'? [2322]

- fargateService.d.ts(169, 5): The expected type comes from property 'taskDefinitionArgs' which is declared here on type 'FargateServiceArgs'
```

However, the actual output is quite cluttered and hard to glean the information from.  This is esp. true in an ide setting where one sees something like this:

image

My suggestion would be to invert the information being provided by errors.  Give the most specific and directly impactful information first (i.e. `Object literal may only specify known properties, but 'enviroment' does not exist in type 'Container'. Did you mean to write 'environment'?`), and then follow that up with all the extra details that can be used to dive deeper into things when that isn't clear enough.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Tries to fix #29759, but I haven't gotten the chance to review the baselines (I know, sounds dumb but I'm on a phone call now). One thing to keep in mind is the fact that we lose context of which branch in a union we're picking. That can be pretty bad, and I'm not sure exactly how to address that.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Looks like this only fails because `jake baseline-accept` doesn't necessarily update unit tests.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12243,6 +12250,7 @@ namespace ts {
                                         reportError(Diagnostics.Object_literal_may_only_specify_known_properties_and_0_does_not_exist_in_type_1,
                                             symbolToString(prop), typeToString(errorTarget));
                                     }
+                                    shouldConstructDiagnosticTrail = false;
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: should probably have explanation?

**Comment by DanielRosenwasser** [comment]
@weswigham @RyanCavanaugh can you take a look?

File: tests/baselines/reference/nestedFreshLiteral.errors.txt [hunk 2]
```diff
@@ -18,9 +15,6 @@ tests/cases/compiler/nestedFreshLiteral.ts(12,21): error TS2322: Type '{ prop: {
     let stylen: NestedCSSProps = {
       nested: { prop: { colour: 'red' } }
                         ~~~~~~~~~~~~~
-!!! error TS2322: Type '{ prop: { colour: string; }; }' is not assignable to type 'NestedSelector'.
-!!! error TS2322:   Types of property 'prop' are incompatible.
-!!! error TS2322:     Type '{ colour: string; }' is not assignable to type 'CSSProps'.
-!!! error TS2322:       Object literal may only specify known properties, but 'colour' does not exist in type 'CSSProps'. Did you mean to write 'color'?
+!!! error TS2561: Object literal may only specify known properties, but 'colour' does not exist in type 'CSSProps'. Did you mean to write 'color'?
 !!! related TS6500 tests/cases/compiler/nestedFreshLiteral.ts:5:3: The expected type comes from property 'nested' which is declared here on type 'NestedCSSProps'
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/nestedFreshLiteral.errors.txt]: Do these two messages mix well together?

**Comment by DanielRosenwasser** [comment]
@sandersn @weswigham any chance you can take a look here?

**Review by RyanCavanaugh — APPROVED** [review]

**Review by weswigham — COMMENTED** [review]
I know I mentioned this the other day, but cutting the entire diagnostic trail seems wrong to me. However, I can certainly see how the `A is not assignable to B` error immediately above the extra props error as usually not being needed. Would it be okay to just suppress a single level of the diagnostic message chain, rather than deleting the entire chain?

Take this one:
```
Conversion of type '{ foo: string; }[]' to type '{ id: number; }[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
Type '{ foo: string; }' is not comparable to type '{ id: number; }'.	
    Object literal may only specify known properties, and 'foo' does not exist in type '{ id: number; }'.
```
cutting it to just 
```
Object literal may only specify known properties, and 'foo' does not exist in type '{ id: number; }'.
```
seems excessive - I have no idea why I'm getting an excess property warning here. But
```
Conversion of type '{ foo: string; }[]' to type '{ id: number; }[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
    Object literal may only specify known properties, and 'foo' does not exist in type '{ id: number; }'.
```
is i think an OK compromise.

**Comment by orta** [comment]
This is /basically/ replaced by https://github.com/microsoft/TypeScript/pull/33473 so I'm closing. I get that they're not exactly the same, but they more or less have the same goals.

**Pull request closed without merging**